weblab/datasets/views.py

```diff
@@ -370,3 +372,110 @@ def get_context_data(self, **kwargs):
+        return self.object
+
+    def get_forms(self):
+        """
+        Construct the forms for this view.
+
+        Populates `self.forms` which is a dict mapping protocol version
+        to a list of forms for that protocol version. Each protocol version
+        has one form per dataset column.
+
+        @return self.forms
+        """
+        dataset = self._get_object()
+
+        self.forms = {}
+
+        # Group existing mappings by protocol version
+        existing_mappings = dataset.column_mappings.all()
+        mappings_by_version = {
+            version: list(mappings)
+            for version, mappings in groupby(existing_mappings, lambda x: x.protocol_version)
+        }
+
+        # Full list of protocol versions for this user
+        protocol_versions = dataset.protocol.cachedentity.versions.visible_to_user(self.request.user)
+
+        for version in protocol_versions:
+            self.forms[version] = []
+
+            # Each form will ist all but FLAG ioputs for this protocol version
+            ioputs = version.ioputs.exclude(kind=(ProtocolIoputs.FLAG))
+
+            # Group existing mappings for this version by column name
+            mappings = {
+                mapping.column_name: mapping
+                for mapping in mappings_by_version.get(version, [])
+            }
+
+            # Create a form for each version/column name combo
+            columns = sorted(dataset.column_names)
+            for (i, column_name) in enumerate(columns):
+                instance = mappings.get(column_name)
+                form = DatasetColumnMappingForm(
+                    prefix='mapping_%d_%d' % (version.pk, i),
+                    data=self.request.POST or None,
+                    instance=instance,
+                    dataset=dataset,
+                    protocol_ioputs=ioputs,
+                    initial={
+                        'dataset': dataset,
+                        'protocol_version': version,
+                    },
+                )
+                self.forms[version].append(form)
+
+        return self.forms
+
+    @property
+    def all_forms(self):
+        """Iterate through all forms in the view"""
+        for _, forms in self.forms.items():
+            yield from forms
+
+    @property
+    def all_forms_valid(self):
+        """Return true if all forms are valid, false otherwise"""
+        return all(form.is_valid() for form in self.all_forms)
+
+    def save_all_forms(self):
+        """Save all forms in the view"""
+        for form in self.all_forms:
+            form.save()
+
+    def get_context_data(self, **kwargs):
+        kwargs['forms'] = self.get_forms()
+        return super().get_context_data(**kwargs)
+
+    def get_success_url(self):
+        """What page to show when the form was processed OK."""
+        dataset = self.object
+        ns = self.request.resolver_match.namespace
+        return reverse(ns + ':map_columns', args=[dataset.id])
```

Comment:
It might be nice to display a message on the returned page so that it's obvious to the user that their settings have been saved. Otherwise they might think nothing happened?

Reply:
I'm not seeing the message appear. Is the necessary JS hook missing in the datasets templates?